File: sdks/go/engine.go
package togglerino

import "encoding/json"

// evaluateFlag evaluates a flag for a given context with segment support.
// This is the SDK-side local evaluation engine, producing results identical
// to the Go backend engine (internal/evaluation/engine.go).
func evaluateFlag(flag FlagDefinition, ctx EvaluationContext, segments []SegmentDefinition) EvaluationResult {
	segMap := buildSegmentMap(segments)

	config := flag.Config

	// 1. If flag is archived, return the fallthrough variant value with reason "archived".
	if flag.Status == "archived" {
		return EvaluationResult{
			Value:   findVariantValue(config.Variants, config.FallthroughVariant, flag.DefaultValue),
			Variant: "",
			Reason:  "archived",
		}
	}

	// 2. If config is disabled, return off variant value with reason "disabled".
	if !config.Enabled {
		return EvaluationResult{
			Value:   findVariantValue(config.Variants, config.OffVariant, flag.DefaultValue),
			Variant: config.OffVariant,
			Reason:  "disabled",
		}
	}

	// 3. Evaluate targeting rules in order (first match wins).
	for _, rule := range config.TargetingRules {
		if matchesAllConditions(rule.Conditions, ctx, segMap) {
			// Check percentage rollout.
			if rule.Percentage != nil {
				bucket := consistentHash(flag.Key, ctx.UserID)
				if bucket >= *rule.Percentage {
					// User is outside the rollout percentage; continue to next rule.
					continue
				}
			}
			// Rule matched — find the variant value.
			value := findVariantValue(config.Variants, rule.Variant, flag.DefaultValue)
			return EvaluationResult{
				Value:   value,
				Variant: rule.Variant,
				Reason:  "rule_match",
			}
		}
	}

	// 4. Return fallthrough variant.
	value := findVariantValue(config.Variants, config.FallthroughVariant, flag.DefaultValue)
	return EvaluationResult{
		Value:   value,
		Variant: config.FallthroughVariant,
		Reason:  "default",
	}
}

// matchesAllConditions checks if all conditions in a rule match the evaluation context.
// When a condition has operator "segment_match", the engine looks up the segment by key
// and evaluates its conditions. Passing nil for segments in the recursive call prevents
// nesting (segments cannot reference other segments).
func matchesAllConditions(conditions []ConditionDefinition, ctx EvaluationContext, segments map[string]SegmentDefinition) bool {
	for _, cond := range conditions {
		if cond.Operator == "segment_match" {
			segKey := cond.Value
			if segments == nil {
				return false
			}
			seg, exists := segments[segKey]
			if !exists {
				return false
			}
			// Evaluate segment conditions (pass nil for segments to prevent nesting).
			if !matchesAllConditions(seg.Conditions, ctx, nil) {
				return false
			}
			continue
		}

		attrValue := getContextValue(cond.Attribute, ctx)
		if !evaluateCondition(cond, attrValue) {
			return false
		}
	}
	return true
}

// getContextValue retrieves a value from the evaluation context.
// Maps "user_id" to ctx.UserID; other attributes come from ctx.Attributes.
func getContextValue(attribute string, ctx EvaluationContext) any {
	if attribute == "user_id" {
		return ctx.UserID
	}
	if ctx.Attributes == nil {
		return nil
	}
	v, ok := ctx.Attributes[attribute]
	if !ok {
		return nil
	}
	return v
}

// findVariantValue finds the value for a variant name in the variants list.
// Returns the deserialized value, or defaultValue if not found.
func findVariantValue(variants []VariantDefinition, name string, defaultValue interface{}) any {
	for _, v := range variants {
		if v.Name == name {
			return rawToAny(v.Value)
		}
	}
	return defaultValue
}

// rawToAny converts json.RawMessage to a Go value.
func rawToAny(raw json.RawMessage) any {
	if raw == nil {
		return nil
	}
	var v any
	if err := json.Unmarshal(raw, &v); err != nil {
		return string(raw)
	}
	return v
}

// buildSegmentMap builds a map from segment key to segment definition for efficient lookup.
func buildSegmentMap(segments []SegmentDefinition) map[string]SegmentDefinition {
	m := make(map[string]SegmentDefinition, len(segments))
	for _, s := range segments {
		m[s.Key] = s
	}
	return m
}

File: sdks/javascript/src/engine.ts
/**
 * Local flag evaluation engine for server-side SDKs.
 * Ports the Go backend's evaluation logic (internal/evaluation/engine.go).
 */

import type { EvaluationContext, EvaluationResult } from './types'
import type {
  FlagDefinition,
  SegmentDefinition,
  ConditionDefinition,
} from './server-types'
import { consistentHash } from './hash'
import { evaluateCondition } from './operators'

/**
 * Evaluates a flag for a given context.
 *
 * @param flag - The flag definition (from the definitions API).
 * @param context - The evaluation context (userId + attributes).
 * @param segments - Array of segment definitions for segment_match resolution.
 * @returns The evaluation result with value, variant, and reason.
 */
export function evaluate(
  flag: FlagDefinition,
  context: EvaluationContext | undefined,
  segments: SegmentDefinition[] | undefined,
): EvaluationResult {
  const ctx = context ?? {}
  const segMap = buildSegmentMap(segments)

  const config = flag.config

  // 1. If flag is archived, return fallthrough variant value with reason "archived".
  if (flag.status === 'archived') {
    return {
      value: lookupVariantValue(flag.variants, config.fallthroughVariant, flag.defaultValue),
      variant: '',
      reason: 'archived',
    }
  }

  // 2. If config is disabled, return off variant value with reason "disabled".
  if (!config.enabled) {
    return {
      value: lookupVariantValue(flag.variants, config.offVariant, flag.defaultValue),
      variant: config.offVariant,
      reason: 'disabled',
    }
  }

  // 3. Evaluate targeting rules in order (first match wins).
  for (const rule of config.targetingRules) {
    if (matchesAllConditions(rule.conditions, ctx, segMap)) {
      // Check percentage rollout.
      if (rule.percentage != null) {
        const bucket = consistentHash(flag.key, ctx.userId ?? '')
        if (bucket >= rule.percentage) {
          // User is outside the rollout percentage; continue to next rule.
          continue
        }
      }
      // Rule matched — find the variant value.
      const value = lookupVariantValue(flag.variants, rule.variant, flag.defaultValue)
      return { value, variant: rule.variant, reason: 'rule_match' }
    }
  }

  // 4. Return fallthrough variant.
  const value = lookupVariantValue(flag.variants, config.fallthroughVariant, flag.defaultValue)
  return { value, variant: config.fallthroughVariant, reason: 'default' }
}

/**
 * Checks if all conditions in a rule match the evaluation context.
 * segment_match conditions look up the segment by key and evaluate its conditions.
 * Passing null for segments in the recursive call prevents nesting.
 */
function matchesAllConditions(
  conditions: ConditionDefinition[],
  ctx: EvaluationContext,
  segments: Map<string, SegmentDefinition> | null,
): boolean {
  for (const cond of conditions) {
    if (cond.operator === 'segment_match') {
      const segKey = cond.value
      if (typeof segKey !== 'string') return false
      if (segments == null) return false
      const seg = segments.get(segKey)
      if (!seg) return false
      // Evaluate segment conditions (pass null for segments to prevent nesting).
      if (!matchesAllConditions(seg.conditions, ctx, null)) return false
      continue
    }

    const attrValue = getContextValue(ctx, cond.attribute)
    const condValue = parseConditionValue(cond.operator, cond.value)
    if (!evaluateCondition(attrValue, cond.operator, condValue)) {
      return false
    }
  }
  return true
}

/**
 * Gets a context value for the given attribute.
 * Maps "user_id" to context.userId; other attributes come from context.attributes.
 */
function getContextValue(ctx: EvaluationContext, attribute: string): unknown {
  if (attribute === 'user_id') {
    return ctx.userId ?? undefined
  }
  return ctx.attributes?.[attribute] ?? undefined
}

/**
 * Parses a condition value for operators that require special handling.
 * For in/not_in operators, the value is a JSON-encoded array string.
 */
function parseConditionValue(operator: string, value: string): unknown {
  if (operator === 'in' || operator === 'not_in') {
    try {
      const parsed = JSON.parse(value)
      if (Array.isArray(parsed)) return parsed
    } catch {
      // Fall through to return raw value.
    }
  }
  return value
}

/**
 * Finds the value for a variant name in the variants list.
 * If the variant is not found, returns the defaultValue.
 */
function lookupVariantValue(
  variants: { name: string; value: unknown }[],
  variantName: string,
  defaultValue: unknown,
): unknown {
  for (const v of variants) {
    if (v.name === variantName) return v.value
  }
  return defaultValue
}

/**
 * Builds a Map<string, SegmentDefinition> from an array for efficient lookup.
 */
function buildSegmentMap(
  segments: SegmentDefinition[] | undefined,
): Map<string, SegmentDefinition> {
  const map = new Map<string, SegmentDefinition>()
  if (segments) {
    for (const s of segments) {
      map.set(s.key, s)
    }
  }
  return map
}
